pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/flink.apache.org/v1beta1/FlinkDeployment/customizations.yaml

```diff
@@ -0,0 +1,110 @@
+apiVersion: config.karmada.io/v1alpha1
+kind: ResourceInterpreterCustomization
+metadata:
+  name: declarative-configuration-flinkdeployment
+spec:
+  target:
+    apiVersion: flink.apache.org/v1beta1
+    kind: FlinkDeployment
+  customizations:
+    healthInterpretation:
+      luaScript: >
+        function InterpretHealth(observedObj)
+          if observedObj.status ~= nil and observedObj.status.jobStatus ~= nil then
+            return observedObj.status.jobStatus.state ~= 'CREATED' and observedObj.status.jobStatus.state ~= 'RECONCILING'
+          end
+          return false
+        end
+    replicaResource:
+      luaScript: >
+        local kube = require("kube")
+
+        local function isempty(s)
+          return s == nil or s == ''
+        end
+
+        function GetReplicas(observedObj)
+          -- FlinkDeployments presently will not be subdivided among clusters, replica should be 1
+          replica = 1
+          requires = {
+            resourceRequest = {},
+          }
+          -- Add jobmanager resources into replica requirement
+
+          jm_replicas = observedObj.spec.jobManager.replicas
+          if isempty(jm_replicas) then
+            jm_replicas = 1
+          end
+
+          for i = 1, jm_replicas do
+            requires.resourceRequest.cpu = kube.resourceAdd(requires.resourceRequest.cpu, tostring(observedObj.spec.jobManager.resource.cpu))
+            requires.resourceRequest.memory = kube.resourceAdd(requires.resourceRequest.memory, observedObj.spec.jobManager.resource.memory)
+          end
+
+          -- Add task manager resources into replica requirement
+
+          parallelism = observedObj.spec.job.parallelism
+          tms = math.ceil(parallelism / observedObj.spec.flinkConfiguration['taskmanager.numberOfTaskSlots'])
+
+          for i = 1, tms do
+            requires.resourceRequest.cpu = kube.resourceAdd(requires.resourceRequest.cpu, tostring(observedObj.spec.taskManager.resource.cpu))
+            requires.resourceRequest.memory = kube.resourceAdd(requires.resourceRequest.memory, observedObj.spec.taskManager.resource.memory)
+          end
+
+          return replica, requires
+        end
```

Comment:
> FlinkDeployments presently will not be subdivided among clusters, replica should be 1

We probably can not just add both `job manager` and `task manager`'s requested resources and treat it just has `one` `super replica`. Otherwise, the karmada-scheduler will try to filter the cluster which has a super node that can hold the the super replica. But that isn't the case, actually, the replicas of `job manager` and `task manager` can be scheduled among different nodes, they do not require a super node to hold all of them.

Reply:
This is a limitation of Karmada for now(there is a plan for this). If a workload(Like FlinkDeployment) has more than one pod template(`JobManager` and `TaskManager`), the resource interpreter can not present each pod template's requirement.

The workaround is that:
1. Sum replicas of `JobManager` and `TaskManager` as the total number of replicas of `FlinkDeployment`
2. When determining the resources for replicas, take the higher resource requirement between JobManager and TaskManager.

Reply:
Ah I see! I'll try this out.

Reply:
This somewhat works, but I would still need to set the replica = 1. Karmada schedules replicas, so if the replica count is greater than 1, Karmada might decide to divide the replicas among clusters and result in duplicate FlinkDeployments on multiple clusters. This is not desired. 

We could keep the replica = 1, and instead set the requirement to be equal to the higher resource requirement between JM and TM as an estimation.

Reply:
Yes, as I said, it is still a workaround before your proposal #5085. But it'll work.

The annoying thing about this workaround approach is it's hard to calculate the required resources.

case 1: If both the JobManager and TaskManger require the same set of resources like they both require CPU and memory, that's easy to figure out which component requires more resources, and then we just need to take higher resource requirements. (lost some accurateness) (but if JobManger requires more CPU and less memory than TaskManager, drop to case 2)

case 2: But what if JobManger and TaskManager require different set of resources, for example, JobManager requires CPU but TaskManager requires memory(or even GPU resources). for this case, draft idea in my mind is that we can take the higher resource requirement of each resource type across both components, something like:
```
          resourceRequest = {}
        
          for resourceType := range taskManager.resource {
              if resourceRequest[resourceType] < taskManager.resource[resourceType] { 
                  resourceRequest[resourceType] = taskManager.resource[resourceType]
              }
          }
        
          for resourceType := range jobManger.resource {
              if resourceRequest[resourceType] < jobManger.resource[resourceType] { 
                  resourceRequest[resourceType] = jobManger.resource[resourceType]
              }
          }
```

Reply:
This is a valid suggestion. It eliminates the false negatives where the cluster has adequate capacity but the scheduler disqualifies it due to the lack of accommodating a super replica.